### PR TITLE
embind: Allow non-ASCII characters in enum values.

### PR DESCRIPTION
**Pull request by marimelon**
Non-ASCII characters cannot be used in enum value.

```C++
// C++
enum Enum { ONE, TWO, Ⅲ, 四 };
```

```JavaScript
// Generated
Type: [Function: ctor] {
      values: [Object],
      ONE: ctor {},
      TWO: ctor {},
      'â\x85¢': ctor {},
      'å\x9B\x9B': ctor {}
}

// Expected
Type: [Function: ctor] {
      values: [Object],
      ONE: ctor {},
      TWO: ctor {},
      'Ⅲ': ctor {},
      '四': ctor {}
}
```

### CODE REVIEW

**Comment by sbc100** [comment]
I've never seen non-ascii characters using as C++ identifiers...  I guess its allowed?

**Comment by marimelon** [comment]
It seems that C++ allows the use of Unicode characters for identifiers.

https://en.cppreference.com/w/cpp/language/identifiers

However, from C++23 it seems to be restricted to UAX31 Default Identifier.

https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1949r6.html
https://github.com/cplusplus/papers/issues/688

**Comment by sbc100** [comment]
Do you have a codebase the uses this?

**Comment by marimelon** [comment]
Sorry, I don't know of any common codebase where non-ascii characters are used in Identifiers.
(Most projects published on Github and elsewhere are written in English for internationalisation purposes.)

I had this problem in a personal project where I used an enum with non-ascii characters. ([repo](https://github.com/marimelon/craftcpp/blob/master/src/Action.hpp))

I also believe that Unicode characters can be used for other identifiers other than enum value. 
In embind, non-ascii characters already seem to be available for function names.

**Comment by sbc100** [comment]
I guess if we go this route we might want consider completely removing `readLatin1String` in favor of UTF8ToString.. but I worry I would be a performance regression when a lot of strings a flowing back and forth. 

Perhaps we could make it an option... and always use one or the other everywhere?

**Comment by marimelon** [comment]
I performed a simple performance test. (nodejs v18.16.1)

```JavaScript
var start1 = process.hrtime()
for (var i = 0; i < 100_000_000; i++) {
  UTF8ToString(name);
}
end1 = process.hrtime(start1)
console.info('Execution time (UTF8ToString): %ds %dms', end1[0], end1[1] / 1000000)

var start2 = process.hrtime()
for (var i = 0; i < 100_000_000; i++) {
  readLatin1String(name);
}
end2 = process.hrtime(start2)
console.info('Execution time (readLatin1String): %ds %dms', end2[0], end2[1] / 1000000)
console.info("name = %s", readLatin1String(name))
```
result
```
Execution time (UTF8ToString): 4s 666.146436ms
Execution time (readLatin1String): 3s 870.925678ms
name = abczyx
```

**Comment by brendandahl** [comment]
For the identifiers that run once performance really won't be an issue, but I don't think we'd want to do UTF8 by default for emval code since that can be very hot with string decodes. I like the idea of one method that is an option that does either latin or utf8.

A note on the benchmark: The result of `readLatin1String` and `UTF8ToString` is not used, so it could potentially be optimized away. You may want to try storing the value and logging it at the end of each loop to ensure it's not removed.

File: src/settings.js [hunk 1]
```diff
@@ -2027,6 +2027,11 @@ var TRUSTED_TYPES = false;
 // settings is *only* needed when also explicitly targeting older browsers.
 var POLYFILL = true;
 
+// Embind specific: If enabled, identifiers are parsed using 'UTF8ToString',
+// allowing the use of Unicode characters. However, there is a slight overhead 
+// during module loading compared to using 'readLatin1String'.
+var ALLOW_UNICODE_IDENTIFIER = false;
```
Comment (brendandahl) [inline src/settings.js:2033]: Let's move this under `EMBIND_STD_STRING_IS_UTF8` in the file and prefix the name e.g. `EMBIND_ALLOW_UNICODE_IDENTIFIER`.

**Review by brendandahl — APPROVED** [review]
LG, just the one comment.

Also, are you planning to add support for putting unicode in other identifiers e.g. 
```c++
function("someUnicode", &foo);
```

**Comment by marimelon** [comment]
> Also, are you planning to add support for putting unicode in other identifiers e.g.

At the moment **I** don't need it.

However, it may be better to be consistent for other identifiers.
If not, it may be necessary to mention this in the setting description.

**Comment by sbc100** [comment]
I'm a little skeptical about whether we want to add this at all.   Every time we add a new setting in emscripten it multiplies our testing burdern.

@marimelon how much does your project really depend on this?  Is it not feasible for you to remove these non-latin enums, or translate/wrap them somehow on the C++ side?

**Comment by marimelon** [comment]
In fact, it is not very important that this PR is merged in my project, as I found that I only needed to apply a few PATCHES to embindwhile creating this PR.

In recent programming languages, it's often possible to use non-Latin characters, which is why I decided to submit this PR.
In reality, since the usage of non-Latin characters is relatively low, I don't mind if this PR is withdrawn.

**Comment by sbc100** [comment]
@marimelon sounds good.   Thanks for looking into this and getting to the bottom of it.   My feeling is that this isn't worth the complexity it adds so I will close this for now.

**Pull request closed without merging**